test/Microsoft.AspNet.Server.KestrelTests/SocketOutputTests.cs

```diff
@@ -199,6 +199,147 @@ public void WritesDontCompleteImmediatelyWhenTooManyBytesIncludingNonImmediateAr
+                Assert.False(task3Success.IsFaulted);
+
+                // Fourth task immediately cancels as the token is cancelled 
+                var task4Throw = socketOutput.WriteAsync(fullBuffer, cancellationToken: cts.Token);
+
+                Assert.True(task4Throw.IsCompleted);
+                Assert.True(task4Throw.IsCanceled);
+                Assert.False(task4Throw.IsFaulted);
+
+                Assert.Throws<TaskCanceledException>(() => task4Throw.GetAwaiter().GetResult());
+
+                // Act
+                while (completeQueue.Count > 0)
+                {
+                    completeQueue.Dequeue()(0);
+                }
+
+                // Give time for tasks to perculate
+                await Task.Delay(2000).ConfigureAwait(false);
+
+                // Second task is cancelled as its token is cancelled 
+                Assert.True(task2Throw.IsCompleted);
+                Assert.True(task2Throw.IsCanceled);
+                Assert.False(task2Throw.IsFaulted);
+
+                Assert.Throws<TaskCanceledException>(() => task2Throw.GetAwaiter().GetResult());
+
+                // Third task is completed and sucessful
+                Assert.True(task3Success.IsCompleted);
+                Assert.False(task3Success.IsCanceled);
+                Assert.False(task3Success.IsFaulted);
+
+                task3Success.GetAwaiter().GetResult();
+
+                var task5Success = socketOutput.WriteAsync(fullBuffer, cancellationToken: default(CancellationToken));
+                // task5 should complete sucessfully as < _maxBytesPreCompleted
+
+                // Fifth task is completed and sucessful
+                Assert.True(task5Success.IsCompleted);
+                Assert.False(task5Success.IsCanceled);
+                Assert.False(task5Success.IsFaulted);
+
+                task5Success.GetAwaiter().GetResult();
+
+                // Sixth task immediately cancels as the token is cancelled 
+                var task6Throw = socketOutput.WriteAsync(fullBuffer, cancellationToken: cts.Token);
+                Assert.True(task6Throw.IsCompleted);
+                Assert.True(task6Throw.IsCanceled);
+                Assert.False(task6Throw.IsFaulted);
+
+                Assert.Throws<TaskCanceledException>(() => task6Throw.GetAwaiter().GetResult());
+
+                var task7Success = socketOutput.WriteAsync(fullBuffer, cancellationToken: default(CancellationToken));
+
+                // Seventh task is not completed
+                Assert.False(task7Success.IsCompleted);
+                Assert.False(task7Success.IsCanceled);
+                Assert.False(task7Success.IsFaulted);
+
+                while (completeQueue.Count > 0)
+                {
+                    completeQueue.Dequeue()(0);
+                }
+
+                await Task.Delay(2000).ConfigureAwait(false);
+
+                // Seventh task is completed and sucessful
+                Assert.True(task7Success.IsCompleted);
```

Comment:
This assertion [failed on AppVeyor](https://github.com/aspnet/KestrelHttpServer/pull/567/files).